src/EditorFeatures/Core/Implementation/CommentSelection/ToggleBlockCommentCommandHandler.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CommentSelection;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.CommentSelection
+{
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.RoslynContentType)]
+    [Name(PredefinedCommandHandlerNames.ToggleBlockComment)]
+    internal class ToggleBlockCommentCommandHandler : AbstractToggleBlockCommentBase
+    {
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        internal ToggleBlockCommentCommandHandler(
+            ITextUndoHistoryRegistry undoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService,
+            ITextStructureNavigatorSelectorService navigatorSelectorService)
+            : base(undoHistoryRegistry, editorOperationsFactoryService, navigatorSelectorService)
+        {
+        }
+
+        /// <summary>
+        /// Gets block comments by parsing the text for comment markers.
+        /// </summary>
+        protected override Task<ImmutableArray<TextSpan>> GetBlockCommentsInDocument(Document document, ITextSnapshot snapshot,
```

Comment:
>GetBlockCommentsInDocument [](start = 58, length = 26)

Is this the impl that will cover JS/TS?  It looks like it's based on string-matching and won't handle things like string literals.  Am I missing something?

Reply:
Interestingly, I notice that VS Code already supports toggle-block-comment for JS/TS.  I'm not sure what LS info they're consuming for that - the only thing I see that would be useful is the span of the enclosing comment.  However, that still requires string matching at the ends of the span since it doesn't indicate the type of the comment.

---
In reply to: [270193035](https://github.com/dotnet/roslyn/pull/34066#discussion_r270193035) [](ancestors = 270193035)

Reply:
@amcasey Currently, all comment handling through roslyn is done with simple textual matching.  This has been, for example, how the current comment/uncomment feature works for all roslyn languages.

The current plan is to proceed with this same approach, with the knowledge that there might be certain cases where legal code is misinterpretted.  So far that hasn't ever been a problem in the past, so we're proceeding with that as a suitable starting point for the new feature as well.  If this is an issue in the future that must be resolved, an appropriate mechanism can be designed to address things.

Reply:
@CyrusNajmabadi that sounds reasonable but then why does C# have it's own syntax tree-based implementation?

Reply:
@dibartbet will have to answer that bit.

Personally, I don't really think it's necessary.

Reply:
VS Code doesn't handle string literals and I haven't heard complaints there either.  Generally speaking, this approach sounds sensible for TS.  Thanks!